ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,305 @@
+        """
+        Returns true if the Agent group uses continuous actions
+        """
+        return self.action_type == ActionType.CONTINUOUS
+
+    @property
+    def action_size(self) -> int:
+        """
+        Returns the dimension of the action.
+         - In the continuous case, will return the number of continuous actions.
+         - In the (multi-)discrete case, will return the number of action.
+         branches.
+        """
+        if self.action_type == ActionType.DISCRETE:
+            return len(self.action_shape)  # type: ignore
+        else:
+            return self.action_shape  # type: ignore
+
+    @property
+    def discrete_action_branches(self) -> Optional[Tuple[int, ...]]:
+        """
+        Returns a Tuple of int corresponding to the number of possible actions
+        for each branch (only for discrete actions). Will return None in
+        for continuous actions.
+        """
+        if self.action_type == ActionType.DISCRETE:
+            return self.action_shape  # type: ignore
+        else:
+            return None
+
+    def create_empty_action(self, n_agents: int) -> np.array:
+        if self.action_type == ActionType.DISCRETE:
+            return np.zeros((n_agents, self.action_size), dtype=np.int32)
+        else:
+            return np.zeros((n_agents, self.action_size), dtype=np.float32)
+
+
+class BaseEnv(ABC):
+    @abstractmethod
+    def step(self) -> None:
+        """
+        Signals the environment that it must move the simulation forward
+        by one step.
+        """
+        pass
+
+    @abstractmethod
+    def reset(self) -> None:
+        """
+        Signals the environment that it must reset the simulation.
+        """
+        pass
+
+    @abstractmethod
+    def close(self) -> None:
+        """
+        Signals the environment that it must close.
+        """
+        pass
+
+    @abstractmethod
+    def get_agent_groups(self) -> List[str]:
+        """
+        Returns the list of the agent group names present in the environment.
+        Agents grouped under the same group name have the same action and
+        observation specs, and are expected to behave similarly in the environment.
+        This list can grow with time as new policies are instantiated.
+        :return: the list of agent group names.
+        """
+        pass
+
+    @abstractmethod
+    def set_actions(self, agent_group: str, action: np.array) -> None:
+        """
+        Sets the action for all of the agents in the simulation for the next
+        step. The Actions must be in the same order as the order received in
+        the step result.
+        :param agent_group: The name of the group the agents are part of
+        :param action: A two dimensional np.array corresponding to the action
+        (either int or float)
+        """
+        pass
+
+    @abstractmethod
+    def set_action_for_agent(
+        self, agent_group: str, agent_id: int, action: np.array
```

Comment:
Do you think it would be worth getting more specific on the types for agent_group and agent_id? Mypy lets you do this: https://mypy.readthedocs.io/en/latest/more_types.html#newtypes
I think it might help prevent some potential errors of using indices as agent IDs.

So something like
```
AgentId = NewType('AgentId', int)
AgentGroup = NewType('AgentGroup', str)
# ...
def set_action_for_agent(
        self, AgentGroup: str, agent_id: AgentId, action: np.array
)
```
Feel free to ignore if you don't think it's helpful.

Reply:
The issue I see is that agent_id are put in a np.array in the batched step result, so we would not be able to return an AgentId as output. If AgentId is an int under the hood, I don't even think Python will be able to catch that if a user used an index instead of an id.
Maybe we can Put a large warning in the docstring ?

Reply:
It wouldn't have any effect on the runtime, just type checking.

Reply:
I will add it then